Chapters 1 & 2 Revisions 05-02-2025 (#5113)
dfinity/motoko: doc/md/fundamentals/2-types/12-advanced-types.md

Structural equality determines whether two values are equal based on their contents. This applies to immutable data structures, such as [records](https://internetcomputer.org/docs/motoko/fundamentals/types/records) and [variants](https://internetcomputer.org/docs/motoko/fundamentals/types/variants), but does not apply to mutable structures for safety reasons.

```motoko
type Point = { x : Int; y : Int };

let p1 : Point = { x = 1; y = 2 };
let p2 : Point = { x = 1; y = 2 };

p1 == p2;  // true (structural equality)
```

> Even though `p1` and `p2` are distinct objects, they are considered equal because they have the same structure and values.

Reviewer: This remains true even if we add some different fields to the point values, since the `==` on Point  values only considers the `x` and `y` fields.

```motoko
type Point = { x : Int; y : Int };
let p1 : Point = { x = 1; y = 2; z = 3 };
let p2 : Point = { x = 1; y = 2; z = 4; c = "Red"; };
p1 == p2;  // true (structural equality at type `Point`)
```

Structural equality determines whether two values are equal based on their contents. This applies to immutable data structures, such as [records](https://internetcomputer.org/docs/motoko/fundamentals/types/records) and [variants](https://internetcomputer.org/docs/motoko/fundamentals/types/variants), but does not apply to mutable structures for safety reasons.

```motoko
type Point = { x : Int; y : Int };

let p1 : Point = { x = 1; y = 2 };
let p2 : Point = { x = 1; y = 2 };

p1 == p2;  // true (structural equality)
```

Even though `p1` and `p2` are distinct objects, they are considered equal because they have the same structure and values.

This remains true even if we add some different fields to the point values, since the `==` on `Point`  values only considers the `x` and `y` fields and ignores other fields.

> ```motoko

Author: <img width="841" alt="Screenshot 2025-05-16 at 2 02 54 PM" src="https://github.com/user-attachments/assets/e119a311-8ec7-4ffd-bb2b-3db186c712d5" />

not sure why its not working for you ?

```motoko
class Box<T>(value : T) {
  public func open() : T { value };
};

let intBox = Box<Nat>(10);
intBox.open();
```

## Recursive types

Recursive types allow a type to refer to itself, enabling the creation of nested structures while maintaining type safety. The base library utilizes recursive types to define linked lists.

```motoko no-repl
type List = ?(Nat, List);
```

This defines a recursive type for representing a linked list of natural number. Each list is either:

- `null`, representing the empty list.

- `?(head, tail)`, where `head` is a `Nat` and `tail` is another `List`.

```motoko
?(1, ?(2, ?(3, null)))  // A list: 1 → 2 → 3
```

To generalize this structure and support values of any type, we introduce a parameterized type:

```motoko no-repl
type List<T> = ?(T, List<T>);
```

This defines a generic linked list, where `T` can be any type (`Nat`, `Text`, `Blob`, or a custom type).

### Manually reversing a linked list

Reversing a linked list involves iterating through the list and prepending each element to a new list. This approach demonstrates list traversal without using `List.reverse` library function.

> Non-paremitized type:

Reviewer: ```suggestion
Non-parameterized type:
```

```motoko

> type List = ?(Nat, List);

Reviewer 2: This example does not run in icp.ninja / dfx

> type List = ?(Nat, List);
> 
> persistent actor NatReverser {
> 
> let numbers : List = ?(1, ?(2, ?(3, null)));
> 
> // Reverses List
> func reverseNat(l : List) : List {
>   var current = l;
>   var rev : List = null;
> 
>   while (current != null) {
>     switch (current) {
>       case (?(head, tail)) {
>         rev := ?(head, rev);
>         current := tail;
>       };
>       case (null) {};
>     };
>   };
>   rev
> };
> 
> reverseNat(numbers); // ?(3, ?(2, ?(1, null)))
> };

Reviewer: ```suggestion
persistent actor NatReverser {

  // Lists of naturals
  type List = ?(Nat, List);

  let numbers : List = ?(1, ?(2, ?(3, null)));

  // Reverses List
  func reverseNat(l : List) : List {
    var current = l;
    var rev : List = null;

    while (current != null) {
      switch (current) {
        case (?(head, tail)) {
          rev := ?(head, rev);
          current := tail;
        };
        case (null) {};
      };
    };
    rev
  };

  reverseNat(numbers); // ?(3, ?(2, ?(1, null)))
};
```

Move type inside and indent. Then it should compile in Ninja.
If you want to display output, add (at top_
```
import Debug "mo:base/Debug"
```
And add bottom do 
```
Debug.print(debug_show reversetNat(numbers));  instead.
```

```

> Parametized:

Reviewer: ```suggestion
Parameterized:
```

Generic types are used to define type parameters that work with multiple data types, commonly used in [functions](https://internetcomputer.org/docs/motoko/fundamentals/types/functions), [classes](https://internetcomputer.org/docs/motoko/fundamentals/types/objects-classes), and data structures.

```motoko
// Generic function
func identity<T>(x : T) : T {
  return x;
};

identity<Nat>(42);  // num is Nat
```

A generic class can store any type while maintaining type safety:

```motoko
class Box<T>(value : T) {
  public func get() : T { value };
};

let intBox = Box<Nat>(10);
intBox.get();
```

> ## Recursive types

Reviewer: This example below (List<T>) actually illustrates two things. recursive types and parameterized types.

Maybe it would be good to first introduce recursive types
List = ?(Nat, List)
And implement  a recursive function reverse.

and then, to introduce parametirized types, generalize to

List<T> = ?(T, List<T>)

and implement generic reverse (using a recursive function).

> Reversing a linked list involves iterating through the list and prepending each element to a new list. This approach demonstrates list traversal and structural mutation without using the built-in `reverse` method.

Reviewer: ```suggestion
Reversing a linked list involves iterating through the list and prepending each element to a new list. This approach demonstrates list traversal without using `List.reverse` library function.
```
The list itself isn't being mutated and can't be. You just some local state to store the new list under construction.

> Type bounds are not limited to records. For example, it is possible to constrain a generic type to be a subtype of a basic type.

Reviewer: ```suggestion
Type bounds are not limited to records. For example, it is possible to constrain a generic type to be a subtype of a basic type.
```
```suggestion
Type bounds are not limited to records. For example, it is possible to constrain a generic type to be a subtype of a primitive type.
```

> ```motoko
> func addIfInt<T <: Int>(x : T, y : T) : Int {
>   return x + y;
> };
> 

Reviewer: ```suggestion
```motoko
func max<T <: Int>(x : T, y : T) : T {
  if (x <= y) y else x
};
```
Is a better example because it now returns a `Nat` when applied to `Nat`, and `Int` when applied to `Int` (the return type depends on T). Otherwise you could just write addIfInt(x : Int; y:Int) : Int which is simpler and has the same utility.

The record example could also be better, for similar reasons.
```

> addIfInt(5, -10);  // Allowed because both are of type Int.

Reviewer: ```suggestion
max<Int>(-5, -10);  // returns -10  : Int
```

```

> 

Reviewer: ````suggestion
But the function can also be used to return the maximum of two `Nat`s and still produce a `Nat` (not an `Int`).
```motoko
max<Nat>(5,10); // returns 5 : Nat
```
````

In the example above, `T <: { name : Text }` requires that any type used for `T` must be a subtype of the [record](https://internetcomputer.org/docs/motoko/fundamentals/types/records) `{ name : Text }`—that is, it must have at least a `name` field of type [`Text`](https://internetcomputer.org/docs/motoko/base/Text). Extra fields are permitted, but the `name` field is mandatory.

Type bounds are not limited to records. For example, it is possible to constrain a generic type to be a subtype of a primitive type.

> ```motoko

Author: works for me its not an error, i suspect the warning is because motoko implicitly casts x and y from type T to Int just for the comparison because it has to assume the type T can be compared using <=

> In the example above, `T <: { name : Text }` requires that any type used for `T` must be a subtype of the [record](https://internetcomputer.org/docs/motoko/fundamentals/types/records) `{ name : Text }`—that is, it must have at least a `name` field of type [`Text`](https://internetcomputer.org/docs/motoko/base/Text). Extra fields are permitted, but the `name` field is mandatory.

Reviewer 2: ```suggestion
In the example above, `T <: { name : Text }` requires that any type used for `T` must be a subtype of the [record](https://internetcomputer.org/docs/motoko/fundamentals/types/records) `{ name : Text }`, that is, it must have at least a `name` field of type [`Text`](https://internetcomputer.org/docs/motoko/base/Text). Extra fields are permitted, but the `name` field is mandatory.
```

Type bounds are not limited to records. For example, it is possible to constrain a generic type to be a subtype of a primitive type.

```motoko name=max
func max<T <: Int>(x : T, y : T) : T {
  if (x <= y) y else x
};
max<Int>(-5, -10);  // returns -5  : Int
```

But the function can also be used to return the maximum of two `Nat`s and still produce a `Nat` (not an `Int`).

```motoko include=max

> max<Nat>(5,10); // returns 10 : Nat

Reviewer: ```suggestion
max<Nat>(5, 10); // returns 10 : Nat
```
